### ISSUE TITLE
Data-driven multilingual filtering

### ISSUE THREAD

**Issue by TomJ-G** [OpenEuroLLM/Taskboard#372]

**Comment by mdocekal** [comment]
I made a sample of HPLT 4 clean (/scratch/project_465002530/training/collection/flag/hplt-4.0/release) consisting of 9 999 585 documents using sqrt weighting upsampling low resource languages.

### Graph showing distribution shift

<img width="1824" height="1050" alt="Image" src="https://github.com/user-attachments/assets/93d3c519-6474-4fab-8912-31948054c51e" />

### Metadata coverage
As not every metadata field is present for every sample, here are the coverage graphs:

<img width="1800" height="1500" alt="Image" src="https://github.com/user-attachments/assets/255ffa71-d0dd-47ed-b143-80050a73b3e3" />

**Comment by mdocekal** [comment]
PCA done on the sample from previous post. Based on @davda54 work.

**Steps:**

1. Z-score normalization
2. PCA
3. Sign convention (orient each PC so its largest-magnitude loading is positive)
4. Components selection (`COMPONENTS_TOPK`) based on Kaiser's rule (eigenvalue > 1)
5. Composite score calculation using following projection matrix `_PROJECTION = (COMPONENTS_TOPK / np.sqrt(EIGENVALUES_TOPK)[:, None]).mean(axis=0)` (weighted sum penalizing too strong components)

### Missing features (metadata fields) problem
There are three strategies for treating missing features

1. `Impute` - Computing normalization without NaNs and setting every NaN to zero afterwards
2. `Drop samples` - drops samples with NaN values
3. `Drop features` - drop features without full coverage

From the results it seems that the `drop features` approach is rather restrictive and the `impute` and `drop samples` behaves more or less the same.

### High level comparison of strategies

<img width="1950" height="750" alt="Image" src="https://github.com/user-attachments/assets/494c26ca-61c1-447f-9320-e5544ede31a9" />

### Scree plots

**Impute**
<img width="1200" height="750" alt="Image" src="https://github.com/user-attachments/assets/72a59b25-cc06-4669-b141-49a4338188fd" />

**Drop samples**
<img width="1200" height="750" alt="Image" src="https://github.com/user-attachments/assets/9acd9a8a-3abf-4448-b719-747d6256410a" />

**Drop features**
<img width="1200" height="750" alt="Image" src="https://github.com/user-attachments/assets/5fee0b45-0271-4cbf-b91b-03865f4fc174" />

### Loadings

**Impute**
<img width="1425" height="1749" alt="Image" src="https://github.com/user-attachments/assets/4d5fad1d-9fde-4849-8c58-cafb998f8a13" />

**Drop samples**
<img width="1425" height="1749" alt="Image" src="https://github.com/user-attachments/assets/aa3e9532-35f6-4f4d-83bc-b981451067c3" />

**Drop features**
<img width="900" height="900" alt="Image" src="https://github.com/user-attachments/assets/a584715b-7ad9-45aa-b0e7-532c6e038906" />

### Pearson correlations

<img width="1848" height="1848" alt="Image" src="https://github.com/user-attachments/assets/51ce52fc-5a2b-4692-a4c7-861ee2a37ee1" />

### Composite scores per language

**Impute**
<img width="2880" height="2940" alt="Image" src="https://github.com/user-attachments/assets/44b7dd23-3c81-4a28-afb6-d3e69614ce8b" />

**Drop samples**
<img width="2880" height="2940" alt="Image" src="https://github.com/user-attachments/assets/e11afd18-0697-4f0d-9716-ea96c0c8b52c" />

**Drop features**
<img width="2880" height="2940" alt="Image" src="https://github.com/user-attachments/assets/a99a7018-0b18-4dc2-b672-44ddd884666c" />

### PCA scatter plots per language for PC1 and PC2

**Impute**
<img width="2880" height="2940" alt="Image" src="https://github.com/user-attachments/assets/fb85cefb-7799-457f-8e0a-ead0f39e625c" />

**Drop samples**
<img width="2880" height="2940" alt="Image" src="https://github.com/user-attachments/assets/81ec54bc-216f-40a0-b039-30b278b964bf" />

**Drop features**
<img width="2880" height="2940" alt="Image" src="https://github.com/user-attachments/assets/c1470d29-d49f-4437-b375-b278f77e8931" />



### Conclusions

1. `Impute` and `Drop samples` strategy bring similar results.
2. `Drop features` strategy is too restrictive.
3. Composite score is not language invariant looking at the [**Composite scores per language**](#composite-scores-per-language) and [**PCA scatter plots per language for PC1 and PC2**](#pca-scatter-plots-per-language-for-pc1-and-pc2).
4. Composite score correlates more with doc_scores and jql (see [**Pearson correlations**](#pearson-correlations)).

**Comment by mdocekal** [comment]
With @TomJ-G we finished the evaluation of this PCA approach.

How it works **TL;DR**:

- **Features:**  precomputed quality signals per document (see obove) (no features computed from raw text)
- **PCA:** features are z-scored and fit with `sklearn.decomposition.PCA`, keeping components with eigenvalue > 1 (Kaiser's rule).
- **Composite score:** each document's projection onto the kept components is reduced to one scalar (using plain mean of the raw PC scores (variance-weighted))
- **Safety filter:** keeping only `"safe"`, `"mild_concerns"`, `"nsfw" `documents — the only hard exclusion in the pipeline.
- **Deciles:** global percentile cutoffs (0–100th, step 10) computed once over the pooled multi-language sample, then applied to bin each language's full corpus into 10 files.
- **Mixture weighting:** each decile file gets a non-zero Megatron blend weight from a softmax over the composite score

**Overall**

<img width="1282" height="819" alt="Image" src="https://github.com/user-attachments/assets/c91fbcb0-94cd-4a56-9d49-e252004ecfa9" />

**Per language**

<img width="3580" height="2133" alt="Image" src="https://github.com/user-attachments/assets/9a712207-5279-4767-bdf8-bd21ea08cd09" />

**Per task**

fin_Latn
<img width="7170" height="2654" alt="Image" src="https://github.com/user-attachments/assets/b774b217-9a38-46c9-82a2-e0d5a7c2ef59" />


hrv_Latn
<img width="7170" height="2654" alt="Image" src="https://github.com/user-attachments/assets/dfaec926-2341-42c8-a54e-381bf660637d" />

por_Latn
<img width="7170" height="3974" alt="Image" src="https://github.com/user-attachments/assets/601f0695-e0d9-42c6-91b2-4f2a8ccfea4d" />

spa_Latn
<img width="7170" height="3974" alt="Image" src="https://github.com/user-attachments/assets/02502afd-9194-44f6-89b5-ad1051689c3f" />

ukr_Cyrl
<img width="7170" height="1334" alt="Image" src="https://github.com/user-attachments/assets/8e4524f7-d33e-4b24-81db-00f916fc44b1" />

**Comment by TomJ-G** [comment]
We have a setback on this tasks: the already filtered data (with sampling approach used for Flag) was selected with PCA approach - the effect of PCA approach on raw, unfiltered data is not known. We are re-running the experiments.
